packages/sign/src/witness/tlog/index.ts

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2023 The Sigstore Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import { encoding as enc } from '../../util';
+import {
+  Entry,
+  ProposedEntry,
+  TLog,
+  TLogClient,
+  TLogClientOptions,
+} from './client';
+import { toProposedEntry } from './entry';
+
+import type { TransparencyLogEntry } from '@sigstore/bundle';
+import type {
+  SignatureBundle,
+  VerificationMaterial,
+  Witness,
+} from '../witness';
+
+export type RekorWitnessOptions = TLogClientOptions;
+
+export class RekorWitness implements Witness {
+  private tlog: TLog;
+
+  constructor(options: RekorWitnessOptions) {
+    this.tlog = new TLogClient(options);
+  }
+
+  public async testify(
+    content: SignatureBundle,
+    publicKey: string
+  ): Promise<VerificationMaterial> {
+    const proposedEntry = toProposedEntry(content, publicKey);
+    const entry = await this.tlog.createEntry(proposedEntry);
+    return toTransparencyLogEntry(entry);
+  }
+}
+
+function toTransparencyLogEntry(entry: Entry): VerificationMaterial {
+  const logID = Buffer.from(entry.logID, 'hex');
+
+  // Parse entry body so we can extract the kind and version.
+  const bodyJSON = enc.base64Decode(entry.body);
+  const entryBody: ProposedEntry = JSON.parse(bodyJSON);
+
+  const promise = entry?.verification?.signedEntryTimestamp
+    ? inclusionPromise(entry.verification.signedEntryTimestamp)
+    : undefined;
+
+  const proof = entry?.verification?.inclusionProof
+    ? inclusionProof(entry.verification.inclusionProof)
+    : undefined;
+
+  const tlogEntry: TransparencyLogEntry = {
+    logIndex: entry.logIndex.toString(),
+    logId: {
+      keyId: logID,
+    },
+    integratedTime: entry.integratedTime.toString(),
+    kindVersion: {
+      kind: entryBody.kind,
+      version: entryBody.apiVersion,
+    },
+    inclusionPromise: promise,
+    inclusionProof: proof,
+    canonicalizedBody: Buffer.from(entry.body, 'base64'),
+  };
+
+  return {
+    tlogEntries: [tlogEntry],
+  };
+}
+
+function inclusionPromise(
+  promise: NonNullable<
+    NonNullable<Entry['verification']>['signedEntryTimestamp']
+  >
+): TransparencyLogEntry['inclusionPromise'] {
+  return {
+    signedEntryTimestamp: Buffer.from(promise, 'base64'),
+  };
+}
+
+function inclusionProof(
+  proof: NonNullable<NonNullable<Entry['verification']>['inclusionProof']>
```

Comment:
The `Entry` type from Rekor gets code-gen'ed as:

```typescript
type Entry {
  verification?: {
    inclusionProof?: InclusionProof | undefined;
  } | undefined;
};
```

For this argument, I just want to the `InclusionProof` type -- it isn't exported from the Rekor package so I have to fish it out of the `Entry` type while working around all of the `undefined` options. Essentially, it boils down to:

> Give me the type of `Entry.verification.inclusionProof` but pretend everything is non-nullable 

Since both Rekor and the Sigstore Bundle define their own (and non-compatible versions) of things like `InclusionProof` and `InclusionPromise` I opted to extract the types from their containing type (either `Entry` or `Bundle`) to provide some context as to which thing we're dealing with. I could export the naked type . . . but then when you see a reference to something like `InclusionProof` it wouldn't be obvious which one you were looking at.